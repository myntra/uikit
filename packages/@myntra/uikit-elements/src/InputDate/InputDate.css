@import '@myntra/tokens/tokens.css';

.gridMonth {
  border-spacing: 0 var(--size-x-small);
}

.gridDay {
  box-sizing: border-box;
  width: var(--size-large);
  height: var(--size-large);
  color: var(--color-black-base);
  text-align: center;
  font-size: var(--size-small);

  /* transition: background 0.2s; */
  line-height: 1;
  cursor: pointer;
}

.gridDayHeader {
  color: rgba(0, 0, 0, 0.38);
  text-align: center;
  font-weight: 500;
  font-size: var(--size-small);
}

.gridToday {
  color: var(--color-secondary);
}

.gridDayDisabled {
  color: rgba(0, 0, 0, 0.27);
  cursor: not-allowed;
}

.gridDaySelected {
  background-color: var(--color-secondary);
  color: white;
}

.gridDayFocused {
  background: rgba(0, 0, 0, 0.1);
}

.gridDayEmpty {
  background-color: transparent;
  cursor: default;
}

.gridDaySelectedStart {
  border-radius: 50% 0 0 50%;
}

.gridDaySelectedEnd {
  border-radius: 0 50% 50% 0;
}

.gridDaySelectedStart.gridDaySelectedEnd {
  border-radius: 50%;
}

.picker {
  display: flex;
  flex-direction: row;
}

.pickerMonth {
  flex-grow: 0;
  flex-shrink: 0;
  background-color: #fff;

  & + & {
    margin-left: var(--size-base-large);
  }
}

/* .pickerMonth + .pickerMonth {
  margin-left: 2px;
} */

.pickerMonthHeader {
  display: flex;
  align-items: center;
  margin-bottom: 8px;
}

.pickerMonthMeta {
  display: flex;
  flex: 1;
  justify-content: center;
  text-align: center;
}

.pickerPrevButton,
.pickerNextButton {
  display: inline-block;
  width: 24px;
  height: 24px;
  opacity: 0.5;
  cursor: pointer;
}

.pickerPrevButton:hover,
.pickerNextButton:hover,
.pickerPrevButton:focus,
.pickerNextButton:focus {
  opacity: 1;
}

.selectWrapper select {
  position: relative;
  display: inline-flex;
  align-items: center;
  padding: 4px 0;
  max-width: 100%;
  outline: none;
  border: none;
  border-radius: 3px;
  background-color: transparent;
  color: var(--color-black-light);
  text-transform: uppercase;
  font-size: 14px;
  line-height: 1.5;
  cursor: pointer;
  -webkit-appearance: none;
  -webkit-box-align: center;
}

.datepicker {
  position: relative;
}

.datepicker.disabled {
  cursor: not-allowed;
}

.datepickerWrapper {
  position: absolute;
  z-index: 1000;
  padding: var(--size-small);
  width: max-content;
  border-radius: 1px;
  background: #fff;
  box-shadow: var(--shadow-light-small);
}

.datepickerControl {
  position: relative;
  display: flex;
  width: 100%;
  background: transparent;
}

.datepickerControlFocus {
  border-bottom-color: #526cd0;
}

.datepickerInput {
  /* flex: 1; */
  background: transparent;
}

.datepickerInput:focus {
  outline: none;
}

.removeIcon {
  position: absolute;
  right: 4px;
  padding: 6px;
  font-size: 12px;
  cursor: pointer;
}

.datePickerControl {
  border-bottom: solid 1px var(--color-gray-base);
}

.datePickerControlFocus {
  border-bottom-color: var(--color-gray-dark);
}

.preset {
  display: flex;
  flex-direction: column;
  margin-left: var(--size-base);
}

.preset_item {
  padding: var(--size-xx-small);
  border-bottom: solid 1px var(--color-gray-base);
  color: var(--color-text-secondary);
  text-transform: capitalize;
  font-size: var(--size-small);
  cursor: pointer;
}

.preset_item_selected {
  background: var(--color-secondary);
  color: #fff;
  cursor: default;
  pointer-events: none;
}
