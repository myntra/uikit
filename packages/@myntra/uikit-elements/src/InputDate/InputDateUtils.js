import PropTypes from 'prop-types'
import { parse as baseParse, format as baseFormat } from 'date-fns'

export const RE_ISO_DATE = /^(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d)$/

export const DateType = PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string])

export const DateRangeType = PropTypes.oneOfType([
  PropTypes.shape({
    from: PropTypes.string,
    to: PropTypes.string
  }),
  PropTypes.shape({
    from: PropTypes.instanceOf(Date),
    to: PropTypes.instanceOf(Date)
  })
])

const padStartWithZeros = (value, digits = 2) => `${value}`.padStart(digits, '0')

/**
 * Create date instance.
 *
 * @export
 * @param {number} year
 * @param {number} month
 * @param {number} date
 * @returns {Date}
 */
export function UTCDate(year, month, date) {
  return new Date(
    `${padStartWithZeros(year, 4)}-${padStartWithZeros(month + 1)}-${padStartWithZeros(date)}T00:00:00.000Z`
  )
}

/**
 * Drop time component from Date.
 *
 * @export
 * @param {Date} date
 * @returns {Date}
 */
export function onlyDate(date) {
  if (typeof date.toDate === 'function') {
    date = date.toDate() // dayjs interop.
  }

  return UTCDate(date.getFullYear(), date.getMonth(), date.getDate())
}

export function parse(str, format) {
  if (str instanceof Date) return str
  if (typeof str !== 'string') return null
  if (!format) throw new Error(`'format' is required.`)
  if (RE_ISO_DATE.test(str)) return new Date(str)
  if (str.length !== format.length) throw new Error(`Date parse exception: unequal length`)

  const date = baseParse(str, format, new Date())

  if (date instanceof Date) {
    return onlyDate(date)
  }

  throw new Error(`Date parse exception: '${str}' does not match '${format}'`)
}

export function format(value, pattern) {
  if (!value) return value
  if (!pattern) return value
  if (value instanceof Date) return baseFormat(value, pattern)
  if (typeof value === 'string') return format(parse(value, pattern), pattern)

  const result = {}
  if (value.from) result.from = format(value.from, pattern)
  if (value.to) result.to = format(value.to, pattern)
  return result
}

export function isDateEqual(a, b) {
  if (!(a instanceof Date)) return false
  if (!(b instanceof Date)) return false
  if (a.getTime() === b.getTime()) return true

  return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate()
}
