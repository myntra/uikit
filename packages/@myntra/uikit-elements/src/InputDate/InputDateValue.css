@import '../Input/Input.css';

.date-value {
  display: inline-flex;
  width: 100%;
}

.date-value > * {
  flex: 1;
}

.wrapper {
  position: relative;
}

.wrapper > .icon {
  cursor: pointer;
  position: absolute;
  right: var(--size-xx-small);
  bottom: var(--size-xx-small);
}

.date-value-active input {
  border-bottom: 1px solid var(--color-primary);
}

.active,
.active input {
  color: var(--color-primary);
}
