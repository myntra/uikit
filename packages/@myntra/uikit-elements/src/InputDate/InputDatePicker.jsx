import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import Picker from './Picker/Picker'
import Month from './Picker/Month'
import Day from './Picker/Day'
import Jumper from './Picker/Jumper'
import { DateType, DateRangeType, parse, format } from './InputDateUtils'
import * as PRESETS from './presets'
import Preset from './Picker/Preset'

const DEFAULT_PRESETS = Object.values(PRESETS)

/**
 *
 * @since 0.0.0
 * @status EXPERIMENTAL
 * @example
 * <InputDatePicker monthsToDisplay={4} value={this.state.value} onChange={value => this.setState({ value })} disabledDates={[{ from: new Date(), to: new Date() }]} />
 */
class InputDatePicker extends PureComponent {
  static propTypes = {
    /**
     * Value
     *
     * @typedef {string|Date} DateType
     *
     * @typedef {{from: DateType, to: DateType}} DateRangeType
     */
    value: PropTypes.oneOfType([DateType, DateRangeType]),
    /**
     * The param type of onChange handle is same the type of value prop.
     *
     * @typedef {string|Date} DateType
     *
     * @typedef {{from: DateType, to: DateType}} DateRangeType
     *
     * @function
     * @param {DateType|DateRangeType} value
     */
    onChange: PropTypes.func,
    presets: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.arrayOf(
        PropTypes.shape({
          range: PropTypes.bool,
          label: PropTypes.string.isRequired,
          /**
           * @typedef {{from: Date, to: Date}} DateRange
           *
           * @function
           * @param {Date} today
           * @returns {Date|DateRange}
           */
          value: PropTypes.func.isRequired
        })
      )
    ]),
    range: PropTypes.bool,
    monthsToDisplay: PropTypes.number,
    /**
     * @typedef {{from: DateType?, to: DateType?}} DateRangeType
     */
    disabledRanges: PropTypes.arrayOf(DateRangeType),
    /**
     * @typedef {string|Date} DateType
     */
    min: DateType,
    /**
     * @typedef {string|Date} DateType
     */
    max: DateType,
    format: PropTypes.string,
    /** @private */
    validate(props) {
      if (
        (props.range
          ? props.value && (typeof props.value.from === 'string' || typeof props.value.to === 'string')
          : typeof props.value === 'string') &&
        !props.format
      ) {
        throw new Error(`Set 'format' prop to 'YYYY-MM-DD' to use string dates.`)
      }
    }
  }

  static defaultProps = {
    range: false,
    disabledRanges: [],
    monthsToDisplay: 1
  }

  handleChange = value => {
    if (this.props.onChange) {
      this.props.onChange(format(value, this.props.format))
    }
  }

  normalize(dateOrDates) {
    // 1. Date or undefined or null
    if (!dateOrDates || dateOrDates instanceof Date) return dateOrDates
    // 2. List of Dates
    if (Array.isArray(dateOrDates)) return dateOrDates.map(date => this.normalize(date))
    // 3. Range object
    if (typeof dateOrDates === 'object' && dateOrDates) {
      const result = {}
      if ('from' in dateOrDates) result.from = this.normalize(dateOrDates.from)
      if ('to' in dateOrDates) result.to = this.normalize(dateOrDates.to)
      return result
    }
    // 4. String!
    if (typeof dateOrDates === 'string') {
      return parse(dateOrDates, this.props.format)
    }
  }

  render() {
    const { value, onChange, disabledRanges, min, max, presets, ...props } = this.props
    const normalizedDate = this.normalize(value)

    return (
      <Picker
        {...props}
        value={normalizedDate}
        onChange={this.handleChange}
        disabledRanges={this.normalize(disabledRanges)}
        min={this.normalize(min)}
        max={this.normalize(max)}
      >
        {this.props.presets && (
          <Preset
            range={this.props.range}
            value={normalizedDate}
            onChange={this.handleChange}
            presets={this.props.presets === true ? DEFAULT_PRESETS : this.props.presets}
          />
        )}
      </Picker>
    )
  }
}

InputDatePicker.Day = Day
InputDatePicker.Month = Month
InputDatePicker.Picker = Picker
InputDatePicker.Jumper = Jumper
InputDatePicker.Preset = Preset

export default InputDatePicker
