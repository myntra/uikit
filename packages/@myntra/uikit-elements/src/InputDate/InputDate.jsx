import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { classnames } from '@myntra/uikit-utils'
import InputDatePicker from './InputDatePicker'
import InputDateValue from './InputDateValue'
import Dropdown from '../Dropdown/Dropdown'
import { format, parse, DateType, DateRangeType, isDateEqual } from './InputDateUtils'
import styles from './InputDate.module.css'

/**
 The InputDate component.

 @since 0.0.0
 @status REVIEWING
 @example
 <div>
 <InputDate displayFormat="MM/dd" value={this.state.value} onChange={(value) => this.setState({value})}/>
 <InputDate displayFormat="MM/dd" value={this.state.value} onChange={(value) => this.setState({value})}/>
 </div>
**/
export default class InputDate extends PureComponent {
  static propTypes = {
    /** @private */
    className: PropTypes.string,
    /** Date format when using string dates. */
    format: PropTypes.string,
    /** Date format used only for displaying date values. */
    displayFormat: PropTypes.string,
    /** Select date range. */
    range: PropTypes.bool,
    /** Disabled */
    disabled: PropTypes.bool,
    /**
     * Change event handler.
     *
     * @typedef {string|Date} DateLike
     *
     * @function
     * @param {DateLike|{ from: DateLike, to: DateLike }} value
     */
    onChange: PropTypes.func,
    /**
     * Value
     *
     * @typedef {string|Date} DateType
     *
     * @typedef {{from: DateType, to: DateType}} DateRangeType
     */
    value: PropTypes.oneOfType([DateType, DateRangeType])
  }

  static defaultProps = {
    disabled: false,
    range: false
  }

  state = {
    isOpen: false,
    openToDate: null,
    activeRangeEnd: null,
    isRangeSelectionActive: false
  }

  constructor(props) {
    super(props)

    this.wrapperRef = element => {
      this.wrapperRef.current = element
    }
  }

  get displayFormat() {
    return this.props.displayFormat || this.props.format || 'yyyy-MM-dd'
  }

  get displayValue() {
    return format(this.props.value, this.displayFormat) || (this.props.range ? {} : '')
  }

  get displayActiveRangeEnd() {
    if (!this.state.isOpen) return null
    if (this.state.activeRangeEnd) return this.state.activeRangeEnd
    if (this.props.value && this.props.value.from && !this.props.value.to) return 'to'
    return 'from'
  }

  get openToDate() {
    if (this.state.openToDate) return this.state.openToDate
    if (!this.props.range) return parse(this.props.value, this.props.format)
    if (!this.props.value) return null
    return parse(
      this.state.activeRangeEnd === 'to'
        ? this.props.value.to || this.props.value.from
        : this.props.value.from || this.props.value.to,
      this.props.format
    )
  }

  handleOpenToDateChange = openToDate => this.setState({ openToDate })

  handleDisplayValueChange = value => {
    if (this.props.onChange) {
      if (this.props.range) {
        value = { ...this.props.value, ...value }

        if (value.from && value.to) {
          const [from, to] = [value.from, value.to].filter(Boolean).sort((a, b) => a.getTime() - b.getTime())

          value = { from, to }
        }
      }
      this.props.onChange(this.props.format ? format(value, this.props.format) : value)
    }
  }

  handleRangeFocus = value => this.setState({ activeRangeEnd: value })
  handleChange = value => {
    let shouldBeClosed = !!(value && !this.props.value)
    if (this.props.range) {
      shouldBeClosed = !!(value && value.from && value.to)
      let nextSelection = null
      if (shouldBeClosed && this.props.value) {
        const hasFromChanged = !isDateEqual(this.props.value.from, value.from)
        const hasToChanged = !isDateEqual(this.props.value.to, value.to)

        // Edited from. Wait for to.
        if (hasFromChanged && !hasToChanged) {
          nextSelection = 'to'
          shouldBeClosed = false
        }
      } else {
        if (!value.to) nextSelection = 'to'
        if (!value.from) nextSelection = 'from'
      }

      this.setState({
        isRangeSelectionActive: !shouldBeClosed,
        activeRangeEnd: nextSelection
      })
    }

    this.props.onChange && this.props.onChange(value)

    if (shouldBeClosed) {
      this.close()
    } else {
      this.skipClose()
    }
  }

  handleDropdownOpen = () => this.setState({ isOpen: true, openToDate: this.openToDate || new Date() })
  handleDropdownClose = () => this.close()

  closeDelayed = () => setTimeout(this.close, 150)

  close = () => {
    if (this._skipThisClose) {
      console.log('Skipping close')
      clearTimeout(this._skipThisClose)
      this._skipThisClose = null

      return
    }

    this.setState({ isOpen: false, activeRangeEnd: null, openToDate: null })
  }

  skipClose = () => {
    this._skipThisClose = setTimeout(() => (this._skipThisClose = null), 150)
  }

  render() {
    return (
      <Dropdown
        auto
        container
        className={classnames(this.props.className, 'input-date').use(styles)}
        isOpen={this.state.isOpen}
        trigger={
          <InputDateValue
            range={this.props.range}
            value={this.displayValue}
            format={this.displayFormat}
            active={this.displayActiveRangeEnd}
            onRangeFocus={this.handleRangeFocus}
            onChange={this.handleDisplayValueChange}
            onBlur={this.closeDelayed}
          />
        }
        onOpen={this.handleDropdownOpen}
        onClose={this.handleDropdownClose}
      >
        <div className={classnames('wrapper').use(styles)} onClick={this.skipClose}>
          <InputDatePicker
            presets={this.props.range}
            disabledDates={[]}
            monthsToDisplay={this.props.range ? 2 : 1}
            {...this.props}
            openToDate={this.openToDate}
            onOpenToDateChange={this.handleOpenToDateChange}
            onChange={this.handleChange}
            active={this.state.activeRangeEnd}
          />
        </div>
      </Dropdown>
    )
  }
}
