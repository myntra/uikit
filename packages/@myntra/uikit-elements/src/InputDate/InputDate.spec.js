import React from 'react'
import { shallow } from 'enzyme'
import InputDate from './InputDate'
import Dropdown from '../Dropdown/Dropdown'
import { UTCDate } from './InputDateUtils'
import InputDatePicker from './InputDatePicker'
import InputDateValue from './InputDateValue'

describe('render', () => {
  it('simple', () => {
    const wrapper = shallow(<InputDate presets={[]} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('range', () => {
    const wrapper = shallow(<InputDate range />)

    expect(wrapper).toMatchSnapshot()
  })
})

const DATE = UTCDate(2018, 4, 5)
describe('date to open', () => {
  test('default to null', () => {
    const wrapper = shallow(<InputDate />)
    const instance = wrapper.instance()
    expect(instance.openToDate).toEqual(null)

    wrapper.find(InputDatePicker).simulate('openToDateChange', DATE)

    expect(instance.openToDate).toEqual(DATE)
  })

  test('value', () => {
    const wrapper = shallow(<InputDate value={DATE} />)
    const instance = wrapper.instance()
    expect(instance.openToDate).toEqual(DATE)

    wrapper.setState({ openToDate: UTCDate(2017, 4, 5) })
    expect(instance.openToDate).toEqual(UTCDate(2017, 4, 5))
  })

  test('date range', () => {
    const wrapper = shallow(<InputDate />)
    const instance = wrapper.instance()
    wrapper.setProps({ range: true, value: { from: DATE } })
    expect(instance.openToDate).toEqual(DATE)

    wrapper.setProps({ range: true, value: { to: DATE } })
    expect(instance.openToDate).toEqual(DATE)

    wrapper.setProps({ range: true, value: { from: DATE } })
    wrapper.setState({ activeRangeEnd: 'to' })
    expect(instance.openToDate).toEqual(DATE)
  })
})

test('active range end', () => {
  const wrapper = shallow(<InputDate range />)
  const instance = wrapper.instance()
  expect(instance.displayActiveRangeEnd).toEqual(null)

  wrapper.setState({ isOpen: true })
  expect(instance.displayActiveRangeEnd).toEqual('from')

  wrapper.setProps({ value: { from: DATE } })
  expect(instance.displayActiveRangeEnd).toEqual('to')

  wrapper.setState({ activeRangeEnd: 'from' })
  expect(instance.displayActiveRangeEnd).toEqual('from')
})

test('display format', () => {
  const wrapper = shallow(<InputDate />)
  const instance = wrapper.instance()
  expect(instance.displayFormat).toEqual('yyyy-MM-dd')

  wrapper.setProps({ format: 'yyyy/MM/dd' })
  expect(instance.displayFormat).toEqual('yyyy/MM/dd')

  wrapper.setProps({ displayFormat: 'yy/MM/dd' })
  expect(instance.displayFormat).toEqual('yy/MM/dd')
})

test('display value', () => {
  const wrapper = shallow(<InputDate />)
  const instance = wrapper.instance()

  expect(instance.displayValue).toEqual('')

  wrapper.setProps({ range: true })
  expect(instance.displayValue).toEqual({})
})

describe('value change', () => {
  test('date', () => {
    const fn = jest.fn()
    const wrapper = shallow(<InputDate onChange={fn} />)

    wrapper.find(InputDatePicker).simulate('change', DATE)

    expect(fn).toHaveBeenCalledWith(DATE)
  })

  test('date range', () => {
    const fn = jest.fn()
    const wrapper = shallow(<InputDate onChange={fn} value={{ from: DATE }} range />)

    wrapper.find(InputDatePicker).simulate('change', { to: DATE })

    expect(fn).toHaveBeenCalledWith({ to: DATE })
  })
})

describe('display value change', () => {
  test('no onChange', () => {
    const wrapper = shallow(<InputDate />)

    expect(() => {
      wrapper
        .find(Dropdown)
        .dive()
        .find(InputDateValue)
        .simulate('change', DATE)
    }).not.toThrow()
  })

  test('date', () => {
    const fn = jest.fn()
    const wrapper = shallow(<InputDate onChange={fn} />)

    wrapper
      .find(Dropdown)
      .dive()
      .find(InputDateValue)
      .simulate('change', DATE)

    expect(fn).toHaveBeenCalledWith(DATE)
  })

  test('date (string)', () => {
    const fn = jest.fn()
    const wrapper = shallow(<InputDate onChange={fn} format="yyyy-MM-dd" />)

    wrapper
      .find(Dropdown)
      .dive()
      .find(InputDateValue)
      .simulate('change', DATE)

    expect(fn).toHaveBeenCalledWith('2018-05-05')
  })

  test('date range', () => {
    const fn = jest.fn()
    const DATE2 = UTCDate(2018, 3, 5)
    const wrapper = shallow(<InputDate onChange={fn} value={{ from: DATE }} range />)

    wrapper
      .find(Dropdown)
      .dive()
      .find(InputDateValue)
      .simulate('change', { to: DATE2 })

    expect(fn).toHaveBeenCalledWith({ from: DATE2, to: DATE })

    wrapper.setProps({ value: {} })
    wrapper
      .find(Dropdown)
      .dive()
      .find(InputDateValue)
      .simulate('change', { to: DATE2 })

    expect(fn).toHaveBeenCalledWith({ to: DATE2 })
  })

  test('date range (string)', () => {
    const fn = jest.fn()
    const wrapper = shallow(<InputDate onChange={fn} value={{ from: DATE }} range format="yyyy-MM-dd" />)

    wrapper
      .find(Dropdown)
      .dive()
      .find(InputDateValue)
      .simulate('change', { to: UTCDate(2018, 3, 5) })

    expect(fn).toHaveBeenCalledWith({ from: '2018-04-05', to: '2018-05-05' })
  })

  test('range focus', () => {
    const wrapper = shallow(<InputDate />)
    wrapper
      .find(Dropdown)
      .dive()
      .find(InputDateValue)
      .simulate('rangeFocus', 'foo')

    expect(wrapper.state().activeRangeEnd).toBe('foo')
  })
})

describe('next selection', () => {
  const wrapper = shallow(<InputDate range />)
  const FROM = UTCDate(2018, 3, 5)
  const TO = UTCDate(2018, 5, 5)

  beforeEach(() => {
    wrapper.setState({
      isOpen: false
    })
  })

  afterEach(() => {
    wrapper.setProps({ value: undefined, range: true })
    wrapper.setState({
      isOpen: false,
      openToDate: null,
      activeRangeEnd: null,
      isRangeSelectionActive: false
    })
  })

  test('select from', () => {
    wrapper.find(InputDatePicker).simulate('change', { from: FROM })

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('to')
  })

  test('select to', () => {
    wrapper.find(InputDatePicker).simulate('change', { to: TO })

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('from')
  })

  test('select none', () => {
    wrapper.find(InputDatePicker).simulate('change', {})

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('from')
  })

  test('select from when to selected', () => {
    wrapper.setProps({ value: { to: TO } })
    wrapper.find(InputDatePicker).simulate('change', { from: FROM })

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('to')
  })

  test('select to when to selected', () => {
    wrapper.setProps({ value: { to: TO } })
    wrapper.find(InputDatePicker).simulate('change', { to: DATE })

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('from')
  })

  test('select both when to selected', () => {
    wrapper.setProps({ value: { to: DATE } })
    wrapper.find(InputDatePicker).simulate('change', { from: FROM, to: TO })

    expect(wrapper.state().isRangeSelectionActive).toBe(false)
    expect(wrapper.state().activeRangeEnd).toBe(null)
  })

  test('select both (swap) when to selected', () => {
    wrapper.setProps({ value: { to: DATE } })
    wrapper.find(InputDatePicker).simulate('change', { from: DATE, to: TO })

    expect(wrapper.state().isRangeSelectionActive).toBe(false)
    expect(wrapper.state().activeRangeEnd).toBe(null)
  })

  test('select from when from selected', () => {
    wrapper.setProps({ value: { from: FROM } })
    wrapper.find(InputDatePicker).simulate('change', { from: DATE })

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('to')
  })

  test('select to when from selected', () => {
    wrapper.setProps({ value: { from: FROM } })
    wrapper.find(InputDatePicker).simulate('change', { to: TO })

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('from')
  })

  test('select both when from selected', () => {
    wrapper.setProps({ value: { from: DATE } })
    wrapper.find(InputDatePicker).simulate('change', { from: FROM, to: TO })

    expect(wrapper.state().isRangeSelectionActive).toBe(false)
    expect(wrapper.state().activeRangeEnd).toBe(null)
  })

  test('select both (swap from) when from selected', () => {
    wrapper.setProps({ value: { from: DATE } })
    wrapper.find(InputDatePicker).simulate('change', { from: FROM, to: DATE })

    expect(wrapper.state().isRangeSelectionActive).toBe(false)
    expect(wrapper.state().activeRangeEnd).toBe(null)
  })

  test('select from when both selected', () => {
    wrapper.setProps({ value: { from: DATE, to: TO } })
    wrapper.find(InputDatePicker).simulate('change', { from: FROM })

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('to')
  })

  test('select to when both selected', () => {
    wrapper.setProps({ value: { from: FROM, to: TO } })
    wrapper.find(InputDatePicker).simulate('change', { to: TO })

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('from')
  })

  test('select both when both selected', () => {
    wrapper.setProps({ value: { from: FROM, to: TO } })
    wrapper.find(InputDatePicker).simulate('change', { from: DATE, to: DATE })

    expect(wrapper.state().isRangeSelectionActive).toBe(false)
    expect(wrapper.state().activeRangeEnd).toBe(null)
  })

  test('select edit from when both selected', () => {
    wrapper.setProps({ value: { from: DATE, to: TO } })
    wrapper.find(InputDatePicker).simulate('change', { from: FROM, to: TO })

    expect(wrapper.state().isRangeSelectionActive).toBe(true)
    expect(wrapper.state().activeRangeEnd).toBe('to')
  })

  test('select edit to when both selected', () => {
    wrapper.setProps({ value: { from: FROM, to: DATE } })
    wrapper.find(InputDatePicker).simulate('change', { from: FROM, to: TO })

    expect(wrapper.state().isRangeSelectionActive).toBe(false)
    expect(wrapper.state().activeRangeEnd).toBe(null)
  })

  test('select both (swap from) when both selected', () => {
    wrapper.setProps({ value: { from: DATE, to: TO } })
    wrapper.find(InputDatePicker).simulate('change', { from: FROM, to: DATE })

    expect(wrapper.state().isRangeSelectionActive).toBe(false)
    expect(wrapper.state().activeRangeEnd).toBe(null)
  })
})

describe('dropdown', () => {
  const wrapper = shallow(<InputDate />)

  wrapper.instance().dropdownRef({
    close() {
      wrapper.find(Dropdown).simulate('close')
    }
  })

  beforeEach(() => {
    wrapper.setState({ isOpen: true })
  })

  describe('open/close', () => {
    const wrapper = shallow(<InputDate />)

    expect(wrapper.state().isOpen).toBe(false)

    wrapper.find(Dropdown).simulate('open')
    expect(wrapper.state().isOpen).toBe(true)

    wrapper.find(Dropdown).simulate('close')
    expect(wrapper.state().isOpen).toBe(false)
  })

  test('force close', () => {
    wrapper.instance().close()
    expect(wrapper.state().isOpen).toBe(false)
  })

  test('close if active element is outside', () => {
    const EVENT_WITH_PATH = { path: [1] }
    const EVENT_WITH_COMPOSED_PATH = { composedPath: () => [1] }
    const EVENT_WITH_CURRENT_TARGET = { currentTarget: { contains: a => [1].includes(a) } }

    wrapper.setState({ isOpen: true })
    wrapper.instance().closeIfElementIsOutsideTarget(EVENT_WITH_PATH, 0)
    expect(wrapper.state().isOpen).toBe(false)

    wrapper.setState({ isOpen: true })
    wrapper.instance().closeIfElementIsOutsideTarget(EVENT_WITH_PATH, 1)
    expect(wrapper.state().isOpen).toBe(true)

    wrapper.setState({ isOpen: true })
    wrapper.instance().closeIfElementIsOutsideTarget(EVENT_WITH_COMPOSED_PATH, 0)
    expect(wrapper.state().isOpen).toBe(false)

    wrapper.setState({ isOpen: true })
    wrapper.instance().closeIfElementIsOutsideTarget(EVENT_WITH_COMPOSED_PATH, 1)
    expect(wrapper.state().isOpen).toBe(true)

    wrapper.setState({ isOpen: true })
    wrapper.instance().closeIfElementIsOutsideTarget(EVENT_WITH_CURRENT_TARGET, 0)
    expect(wrapper.state().isOpen).toBe(false)

    wrapper.setState({ isOpen: true })
    wrapper.instance().closeIfElementIsOutsideTarget(EVENT_WITH_CURRENT_TARGET, 1)
    expect(wrapper.state().isOpen).toBe(true)
  })
})
