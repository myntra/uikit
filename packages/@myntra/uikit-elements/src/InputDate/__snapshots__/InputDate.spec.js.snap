// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render range 1`] = `
<Dropdown
  approxContentHeight={320}
  approxContentWidth={240}
  auto={true}
  className="input-date"
  onBlur={[Function]}
  onClose={[Function]}
  onOpen={[Function]}
  trigger={
    <InputDateValue
      active={null}
      format="yyyy-MM-dd"
      onChange={[Function]}
      onRangeFocus={[Function]}
      range={true}
      value={Object {}}
    />
  }
>
  <div
    className="wrapper"
  >
    <InputDatePicker
      active={null}
      disabled={false}
      disabledDates={Array []}
      disabledRanges={Array []}
      monthsToDisplay={2}
      onChange={[Function]}
      onOpenToDateChange={[Function]}
      openToDate={null}
      presets={true}
      range={true}
    />
  </div>
</Dropdown>
`;

exports[`render simple 1`] = `
<Dropdown
  approxContentHeight={320}
  approxContentWidth={240}
  auto={true}
  className="input-date"
  onBlur={[Function]}
  onClose={[Function]}
  onOpen={[Function]}
  trigger={
    <InputDateValue
      active={null}
      format="yyyy-MM-dd"
      onChange={[Function]}
      onRangeFocus={[Function]}
      range={false}
      value=""
    />
  }
>
  <div
    className="wrapper"
  >
    <InputDatePicker
      active={null}
      disabled={false}
      disabledDates={Array []}
      disabledRanges={Array []}
      monthsToDisplay={1}
      onChange={[Function]}
      onOpenToDateChange={[Function]}
      openToDate={null}
      presets={Array []}
      range={false}
    />
  </div>
</Dropdown>
`;
